python/ray/autoscaler/_private/vsphere/node_provider.py

```diff
@@ -0,0 +1,726 @@
+                        nodes.append(vm_id)
+                        # refresh cached_nodes with latest information e.g external ip
+                        self.cached_nodes[vm_id] = vm
+
+            cli_logger.info(f"Nodes are {nodes}")
+            return nodes
+
+    def is_running(self, node_id):
+        node = self._get_cached_node(node_id)
+        return node.power_state in {HardPower.State.POWERED_ON}
+
+    def is_terminated(self, node_id):
+        node = self._get_cached_node(node_id)
+        return node.power_state not in {HardPower.State.POWERED_ON}
+
+    def node_tags(self, node_id):
+        with self.tag_cache_lock:
+            d1 = self.tag_cache[node_id]
+            return dict(d1)
+
+    def external_ip(self, node_id):
+        # Return the external IP of the VM
+
+        vm = self.vsphere_sdk_client.vcenter.vm.guest.Identity.get(node_id)
+        try:
+            _ = ipaddress.IPv4Address(vm.ip_address)
+            cli_logger.info("Fetch IP {} for VM {}".format(vm.ip_address, vm))
+        except ipaddress.AddressValueError:
+            # vSphere SDK could return IPv6 address when the VM is just booted. We
+            # just return None in this case because the Ray doesn't support IPv6
+            # address yet When the next time external_ip is called, we could return
+            # the IPv4 address
+            return None
+        return vm.ip_address
+
+    def internal_ip(self, node_id):
+        # Currently vSphere VMs do not show an internal IP. So we just return the
+        # external IP
+        return self.external_ip(node_id)
+
+    def set_node_tags(self, node_id, tags):
+
+        cli_logger.info("Setting tags for vm {}".format(node_id))
+        # This method gets called from the Ray and it passes
+        # node_id which needs to be vm.vm and not vm.name
+        with self.lock:
+            category_id = self.get_category()
+            if not category_id:
+                category_id = self.create_category()
+
+            for key, value in tags.items():
+
+                tag = kv_pair_to_vsphere_tag(key, value)
+                tag_id = self.get_tag(tag, category_id)
+                if not tag_id:
+                    tag_id = self.create_node_tag(tag, category_id)
+
+                # If a tag with a key is present on the VM, then remove it
+                # before updating the key with a new value.
+                self.remove_tag_from_vm(key, node_id)
+
+                cli_logger.info(f"Attaching tag {tag} to {node_id}")
+                self.attach_tag(node_id, Constants.TYPE_OF_RESOURCE, tag_id=tag_id)
+
+    def create_node(self, node_config, tags, count) -> Dict[str, Any]:
+        """Creates instances.
+
+        Returns dict mapping instance id to VM object for the created
+        instances.
+        """
+        filters = tags.copy()
+        if TAG_RAY_CLUSTER_NAME not in tags:
+            filters[TAG_RAY_CLUSTER_NAME] = self.cluster_name
+
+        counter = count
```

Comment:
After reading the code, it looks like this variable is meant to represent the number of nodes that need to be started from scratch, which the `count` minus the number of reused nodes.  Might be worth adding a comment for readability.

Reply:
I think a meaningful variable is also a straighforward and good way for providing that information, changed that.